File: SwiftTracer/SwiftTracer/Scene.swift
import Foundation
import Cocoa
import AppKit


class Scene {
    let objects : [SceneObject]
    let lights : [Light]
    let camera : Camera
    
    init(objects:[SceneObject], lights:[Light], camera : Camera) {
        self.objects = objects
        self.lights = lights
        self.camera = camera
    }
    
    func intersection(ray:Ray) -> Optional<Intersection> {
        var intersection2 : Optional<Intersection> = .None
        for object in self.objects {
            let isect = object.Intersects(ray)
            switch (isect, intersection2) {
            case (let .Some(i1), let .Some(i2)) where i1.distance < i2.distance : intersection2 = i1
            case (let .Some(_), .None) : intersection2 = isect
            case (_, let i) : intersection2 = i
            }
        }
        return intersection2
    }
    
    func testRay(ray:Ray) -> Double {
        let intersection = self.intersection(ray)
        switch intersection {
        case let .Some(intersection) : return intersection.distance
        case let .None : return 0.0
        }
    }
    
    func traceRay(ray:Ray, depth:Int) -> Color {
        let black = Color(r:0.0, g:0.0, b:0.0)
        let isect = self.intersection(ray)
        switch isect {
        case .Some(let isect) : return self.shade(isect, depth: depth)
        case .None : return black
        }
    }
    
    func shade(intersection:Intersection, depth:Int) -> Color {
        let pos = intersection.ray.direction.Scale(intersection.distance).Add(intersection.ray.position)
        let normal = intersection.object.Normal(pos)
        let dot = normal.Dot(intersection.ray.direction)
        let reflectDir = intersection.ray.direction.Subtract(normal.Scale (2.0 * dot))
        let natural = self.getNaturalColor(intersection.object, position: pos, normal: normal, rayDirection: reflectDir)
        if depth < 5 {
            let reflectPos = reflectDir.Scale(0.001).Add(pos)
            let reflectColor : Color = self.getReflectionColor (intersection.object, position: reflectPos, rayDirection: reflectDir, depth: depth)
            return reflectColor.Add(natural)
        } else {
            return Color(r:0.5,g:0.5,b:0.5).Add(natural)
        }
    }
    
    func getNaturalColor(sceneObject : SceneObject, position : Vector, normal : Vector, rayDirection : Vector) -> Color {
        let black = Color(r:0.0, g:0.0, b:0.0)
        var result = black
        for light in scene.lights {
            let lightDistance = light.position.Subtract(position)
            let livec = lightDistance.Normalize()
            let ray = Ray(position: position, direction: livec)
            let isect = scene.testRay(ray)
            let isInShadow = (isect > lightDistance.Length() || isect == 0.0)
            if isInShadow {
                let illum = livec.Dot(normal)
                
                var lcolor = black
                if illum > 0.0 {
                    lcolor = light.color.Scale(illum)
                }
                let normalizedRayDirection = rayDirection.Normalize()
                let specular = livec.Dot(normalizedRayDirection)
                var scolor = black
                if specular > 0.0 {
                    scolor = light.color.Scale(pow(specular, sceneObject.GetSurface().Roughness()))
                }
                
                let diffuseResult = sceneObject.GetSurface().Diffuse(position).Multiply(lcolor)
                let specularResult = sceneObject.GetSurface().Specular(position).Multiply(scolor)
                
                result = result.Add(diffuseResult).Add(specularResult)
            }
        }
        return result
    }
    
    func getReflectionColor(sceneObject : SceneObject, position : Vector, rayDirection : Vector, depth : Int) -> Color {
        let ray = Ray(position: position, direction: rayDirection)
        return (traceRay(ray, depth: depth + 1).Scale(sceneObject.GetSurface().Reflect(position)))
    }
    
    func recenterX (x:Int, halfWidth:Double, invWidth:Double) -> Double {
        let x = Double(x)
        return (x - halfWidth) * invWidth
    }
    
    func recenterY (y:Int, halfHeight:Double, invHeight:Double) -> Double {
        let y = Double(y)
        return -(y - halfHeight) * invHeight
    }
    
    func getPoint(x:Int, y:Int, halfWidth:Double, halfHeight:Double, invWidth:Double, invHeight:Double, camera : Camera) -> Vector {
        let right = camera.right.Scale(recenterX(x, halfWidth: halfWidth, invWidth: invWidth))
        let up = camera.up.Scale(recenterY(y, halfHeight: halfHeight, invHeight: invHeight))
        return right.Add(up).Add(camera.forward).Normalize()
    }
    
    func Render(width:Int, height:Int) -> [NSColor] {
        let fWidth = Double(width)
        let fHeight = Double(height)
        let inverseAspectRatio = fHeight / fWidth
        let scaleWidth = fWidth * inverseAspectRatio
        let halfWidth = scaleWidth / 2.0
        let halfHeight = fHeight / 2.0
        let invWidth = 1.0 / (scaleWidth * 2.0)
        let invHeight = 1.0 / (fHeight * 2.0)
        var result = Array<NSColor>(count:width * height, repeatedValue:NSColor())
        var basey = 0
        for y in 0 ... height - 1 {
            basey = y * width
            for x in  0 ... width - 1 {
                let rayDir = getPoint(x, y: y, halfWidth: halfWidth, halfHeight: halfHeight, invWidth: invWidth, invHeight: invHeight, camera: self.camera)
                let ray = Ray(position: self.camera.position, direction: rayDir)
                result[x + basey] = self.traceRay(ray, depth:0).ToColor()
            }
        }
        return result
    }
}

File: CPlusPlusTracer/CPlusPlusTracer/Scene.cpp
#include "Scene.h"

CScene::CScene(CCamera camera,
   vector<CSceneObject*> objects,
   vector<CLight> lights)
   : Camera(camera), Objects(objects), Lights(lights) {
}

shared_ptr<CIntersection> CScene::Intersection(CRay ray) {
   shared_ptr<CIntersection> intersection = nullptr;
   for (int i = 0; i < this->Objects.size(); i++) {
      auto object = this->Objects[i];
      auto isect = object->Intersects(ray);
      if (isect != nullptr) {
         if (intersection == nullptr) {
            intersection = isect;
         }
         else if (isect->GetDistance() < intersection->GetDistance()) {
            intersection = isect;
         }
      }
   }
   return intersection;
}

double CScene::TestRay(CRay ray) {
   auto intersection = this->Intersection(ray);
   return intersection == nullptr ? 0.0 : intersection->GetDistance();
}

CColor CScene::TraceRay(CRay ray, int depth) {
   auto black = CColor(0.0, 0.0, 0.0);
   auto intersection = this->Intersection(ray);
   return intersection == nullptr ? black : this->Shade(intersection, depth);
}

CColor CScene::Shade(shared_ptr<CIntersection> intersection, int depth) {
   auto pos = intersection->GetRay().GetDirection().Scale(intersection->GetDistance()).Add(intersection->GetRay().GetPosition());
   auto normal = intersection->GetObject()->Normal(pos);
   auto dot = normal.Dot(intersection->GetRay().GetDirection());
   auto reflectDir = intersection->GetRay().GetDirection().Subtract(normal.Scale(2.0 * dot));
   auto natural = this->GetNaturalColor(intersection->GetObject(), pos, normal, reflectDir);
   if (depth < 5) {
      auto reflectPos = reflectDir.Scale(0.0001).Add(pos);
      auto reflectColor = this->GetReflectionColor(intersection->GetObject(), reflectPos, reflectDir, depth);
      return natural.Add(reflectColor);
   }
   auto grey = CColor(0.5, 0.5, 0.5);
   return grey.Add(natural);
}

CColor CScene::GetNaturalColor(CSceneObject* object, CVector position, CVector normal, CVector rayDirection) {
   auto black = CColor(0.0, 0.0, 0.0);
   auto result = black;
   for (int i = 0; i < this->Lights.size(); i++) {
      auto light = this->Lights[i];
      auto lightDistance = light.GetPosition().Subtract(position);
      auto livec = lightDistance.Normalize();
      auto ray = CRay(position, livec);
      auto isect = this->TestRay(ray);
      auto isInShadow = (isect > lightDistance.Length() || isect == 0.0);
      if (isInShadow) {
         auto illum = livec.Dot(normal);
         auto lcolor = illum > 0 ? light.GetColor().Scale(illum) : black;
         auto normalizedRayDirection = rayDirection.Normalize();
         auto specular = livec.Dot(normalizedRayDirection);
         auto scolor = specular > 0.0 ? light.GetColor().Scale(pow(specular, object->GetSurface()->Roughness())):black;
         auto diffuseResult = object->GetSurface()->Diffuse(position).Multiply(lcolor);
         auto specularResult = object->GetSurface()->Specular(position).Multiply(scolor);
         result = result.Add(diffuseResult).Add(specularResult);
      }
   }

   return result;
}

CColor CScene::GetReflectionColor(CSceneObject* object, CVector pos, CVector rayDirection, int depth) {
   auto ray = CRay(pos, rayDirection);
   return this->TraceRay(ray, depth + 1).Scale(object->GetSurface()->Reflect(pos));
}

double CScene::RecenterX(int x, double halfWidth, double invWidth) {
   double fx = (double)x;
   return (fx - halfWidth) * invWidth;
}

double CScene::RecenterY(int y, double halfHeight, double invHeight) {
   double fy = (double)y;
   return -(fy - halfHeight) * invHeight;
}

CVector CScene::GetPoint(int x, int y, double halfWidth, double halfHeight, double invWidth, double invHeight, CCamera camera) {
   auto right = camera.GetRight().Scale(this->RecenterX(x, halfWidth, invWidth));
   auto up = camera.GetUp().Scale(this->RecenterY(y, halfHeight, invHeight));
   return right.Add(up).Add(camera.GetForward()).Normalize();
}

int* CScene::Render(int width, int height, int* result) {
   auto fWidth = (double)width;
   auto fHeight = (double)height;
   auto inverseAspectRatio = fHeight / fWidth;
   auto scaleWidth = fWidth * inverseAspectRatio;
   auto halfWidth = scaleWidth / 2.0;
   auto halfHeight = fHeight / 2.0;
   auto invWidth = 1.0 / (scaleWidth * 2.0);
   auto invHeight = 1.0 / (fHeight * 2.0);

   int basey = 0;
   for (int y = 0; y < height; y++) {
      basey = y * width;
      for (int x = 0; x < width; x++) {
         auto rayDir = this->GetPoint(x, y, halfWidth, halfHeight, invWidth, invHeight, this->Camera);
         auto ray = CRay(this->Camera.GetPosition(), rayDir);
         result[basey + x] = this->TraceRay(ray, 0).ToColor();
      }
   }
   return result;
}